cmd/record_test.go

```diff
@@ -55,48 +55,99 @@ func reqString(t *testing.T, req *http.Request) string {
 }
 
 func TestRecord(t *testing.T) {
-	f := testfile(t, "hook.yml")
-	defer deletefile(t, f)
-
-	r, err := newRecorder(f.Name())
-	if err != nil {
-		t.Fatal(err)
-	}
-	defer r.close()
-
-	srv := httptest.NewServer(r)
-	defer srv.Close()
-
-	req, err := http.NewRequest(http.MethodGet, srv.URL, strings.NewReader(""))
-	if err != nil {
-		t.Fatal(err)
-	}
-	s := `method: GET
+	testcases := []struct {
+		name    string
+		want    string
+		method  string
+		headers http.Header
+		body    string
+		query   string
+	}{
+		{
+			name: "empty body and params",
+			want: `method: GET
 headers:
   Accept-Encoding:
   - gzip
   User-Agent:
   - Go-http-client/1.1
 body: ""
-`
-
-	client := http.DefaultClient
-
-	if _, err := client.Do(req); err != nil {
-		t.Fatal(err)
-	}
-	got := readfile(t, f)
-	if d := diff.Diff(s, got); d != "" {
-		t.Error(d)
+params: {}
+`,
+			method: http.MethodGet,
+			body:   "",
+			query:  "",
+		},
+		{
+			name: "body and params with headers",
+			want: `method: POST
+headers:
+  Accept-Encoding:
+  - gzip
+  Captain:
+  - Hook
+  Content-Length:
+  - "5"
+  User-Agent:
+  - Go-http-client/1.1
+body: tacos
+params:
+  key:
+  - value
+  other:
+  - one
+  - two
+`,
+			method:  http.MethodPost,
+			headers: http.Header{"Captain": {"Hook"}},
+			body:    "tacos",
+			query:   "?key=value&other=one&other=two",
+		},
 	}
 
-	// Make request again to test appends.
-	if _, err := client.Do(req); err != nil {
-		t.Fatal(err)
-	}
-	want := fmt.Sprintf("%s---\n%s", s, s)
-	got = readfile(t, f)
-	if d := diff.Diff(want, got); d != "" {
-		t.Error(d)
+	for _, tc := range testcases {
+		f := testfile(t, "hook.yml")
+
+		r, err := newRecorder(f.Name())
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		srv := httptest.NewServer(r)
+
+		req, err := http.NewRequest(tc.method, srv.URL+tc.query, strings.NewReader(tc.body))
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		for k, v := range tc.headers {
```

Comment:
I think you can replace with with `req.Header = tc.headers`

Reply:
Good catch. I was worried that would blow away the default headers but I it makes sense that they are populated last.